### PR TITLE
Features/227 lshape

### PR DESCRIPTION
**Pull request by ClaudiaComito**
@Markus-Goetz there's still a quirk here with argmin and argmax. Even if axis = 0 (or contains 0), we can't lose the first dimension of partial, because partial_op in this case yields both the values and the indices, so dimension 0 of partial is by default at least 2 when it gets returned to the function. The fix has to happen in the argmin()/ argmax() functions. Not sure how to do that (yet).

Example:

```
>>> a = ht.random.randn(3,4,5)
>>> a
tensor([[[-0.3470, -0.3984,  0.0270,  0.3000,  1.0776],
         [-1.2756, -0.8707,  0.8296, -0.7750, -1.5317],
         [-0.2380, -0.3360, -1.4865,  0.5521, -0.2564],
         [-0.6830, -0.3591, -1.3159,  0.5883,  1.9217]],

        [[ 0.4855, -1.9448,  0.6220,  1.9253, -0.8722],
         [ 0.3382, -0.5159,  0.5053,  1.6421, -1.1279],
         [-0.5231, -0.1203,  1.3909,  0.4133,  0.6540],
         [-1.3037, -1.0900, -0.3871, -0.7963,  0.9564]],

        [[-1.0519,  0.0965, -0.9878, -0.0702, -0.2331],
         [ 0.0111, -1.1165, -1.6232, -0.1156, -0.3195],
         [-0.1528,  0.5501,  0.2335,  0.1948,  0.7384],
         [-1.8235,  1.4258,  0.2019,  1.0923,  1.0515]]])
>>> a.argmin(0)
tensor([[[2, 1, 2, 2, 1],
         [0, 2, 2, 0, 0],
         [1, 0, 0, 2, 0],
         [2, 1, 0, 1, 1]]])
>>> a.argmin(0).shape
(4, 5)
>>> a.argmin(0).lshape
(1, 4, 5)

```

With axis = 1:

```
>>> a.argmin(1)
tensor([[1, 1, 2, 1, 1],
        [3, 0, 3, 3, 1],
        [3, 1, 1, 1, 1]])
>>> a.argmin(1).shape
(3, 5)
>>> a.argmin(1).lshape
(3, 5)
```

### CODE REVIEW

**Comment by Markus-Goetz** [comment]
Could the fix be a squeeze() after the partial call?

**Comment by ClaudiaComito** [comment]
> Could the fix be a squeeze() after the partial call?

Thanks, indeed, that's probably what I need. Would ht.squeeze fit with the reductions or do we need to inaugurate the array_manipulations subgroup of functions?

File: heat/core/operations.py [hunk 1]
```diff
@@ -731,11 +855,20 @@ def __reduce_op(x, partial_op, reduction_op, **kwargs):
...
+                output_shape = output_shape[:dim] + (1,) + output_shape[dim + 1:]
+        if not keepdim:
+            gshape_losedim = tuple(x.gshape[dim] for dim in range(len(x.gshape)) if not dim in axis)
+            lshape_losedim = tuple(x.lshape[dim] for dim in range(len(x.lshape)) if not dim in axis)
+            output_shape = gshape_losedim
+            if reduction_op in [MPI_ARGMAX, MPI_ARGMIN]:
+                if 0 in axis:
+                    lshape_losedim = (2,) + lshape_losedim
+                else:
+                    lshape_losedim = (2 * lshape_losedim[0],) + lshape_losedim[1:]
```
Comment (coquelin77) [inline heat/core/operations.py]: the 2 seems fishy to me. can you justify it for me? I am just not seeing it

File: heat/core/tensor.py [hunk 2]
```diff
@@ -278,7 +280,7 @@ def allclose(self, other, rtol=1e-05, atol=1e-08, equal_nan=False):
         """
         return operations.allclose(self, other, rtol, atol, equal_nan)
 
-    def any(self, axis=None, out=None):
+    def any(self, axis=None, out=None, keepdim=None):
```
Comment (coquelin77) [inline heat/core/tensor.py]: keepdim=False vs =None again

File: heat/core/tensor.py [hunk 3]
```diff
@@ -1360,6 +1362,55 @@ def sqrt(self, out=None):
...
+    def squeeze(self, axis=None):
+        """
+        Remove single-dimensional entries from the shape of a tensor.
+
+        Parameters:
+        -----------
+        x : ht.tensor
+            Input data.
+
+        axis : None or int or tuple of ints, optional
```
Comment (coquelin77) [inline heat/core/tensor.py]: more explanation for axis as mentioned before

File: heat/core/operations.py (hunk 1 above)
Comment (ClaudiaComito) [reply heat/core/operations.py]: Hi Daniel @coquelin77, 
thanks, I'm really unhappy with this if statement. The problem is that, if axis is not None,  the result of local_argmin/max and MPI_ARGMIN/MAX is made up of two sets of results really: the min/max values along a dimension, and the respective indices. So at this stage the dimension along the reduced axis isn't 1 (like for every other reduction operation), it's 2. The min/max values get stripped out of the result at a later stage in the argmin/max() functions.

It's cumbersome though and giving me a lot of trouble pretty much whatever I'm working on. So @Markus-Goetz, @coquelin77  I'm thinking of  adding  __argreduce_op to operations.py and funnel argmin and argmax out to that function, leaving __reduce_op for "regular" reduction operations.

File: heat/core/operations.py [hunk 4]
```diff
@@ -208,8 +209,7 @@ def allclose(x, y, rtol=1e-05, atol=1e-08, equal_nan=False):
     return bool(_local_allclose.item())
 
 
-
-def any(x, axis=None, out=None):
+def any(x, axis=None, out=None, keepdim=None):
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/operations.py]: Done (now in logical.py).

File: heat/core/operations.py [hunk 5]
```diff
@@ -466,6 +493,102 @@ def copy(a):
...
+def squeeze(x, axis=None):
+    """
+    Remove single-dimensional entries from the shape of a tensor.
+
+    Parameters:
+    -----------
+    x : ht.tensor
+        Input data.
+
+    axis : None or int or tuple of ints, optional
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/operations.py]: Done (now in manipulations.py)

File: heat/core/operations.py (hunk 1 above)
Comment (ClaudiaComito) [reply heat/core/operations.py]: OK here's what it looks like now (operations.py lines 217-221)
```
# Take care of special cases argmin and argmax: keep partial.shape[0]
            if (0 in axis and partial.shape[0] != 1):
                lshape_losedim = (partial.shape[0],) + lshape_losedim
            if (not 0 in axis and partial.shape[0] != x.lshape[0]):
                lshape_losedim = (partial.shape[0],) + lshape_losedim[1:]
```
Basically, the assumption is that whenever the first dimension of partial is different from what it should be (1 if reduction along axis 0, or x.lshape[0] if reduction along any other axis), there will be a good reason for it so just keep that partial[0] as first dimension.

This way we don't need to add ifs for every quirky reduction operation that comes our way.

Please let me know if I'm overseeing something.

File: heat/core/tensor.py (hunk 2 above)
Comment (ClaudiaComito) [reply heat/core/tensor.py]: done (dndarray.py)

File: heat/core/tensor.py (hunk 3 above)
Comment (ClaudiaComito) [reply heat/core/tensor.py]: done (dndarray.py)

**Comment by Markus-Goetz** [comment]
Bump, has a minor conflict with the master

**Comment by ClaudiaComito** [comment]
Alright @Markus-Goetz , @coquelin77 , I've commented out the test that fails for now, fixing that (Issue #273) needs fixing Allgatherv (#233, @Cdebus ). It would still be good to have squeeze() and the lshape fixes merged into master.

Thanks,

  Claudia

**Pull request merged**